src/hammer-vlsi/hammer_vlsi/submit_command.py

```diff
@@ -0,0 +1,236 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+# submit_command.py
+#
+
+from abc import ABCMeta, abstractmethod
+import atexit
+import subprocess
+import hammer_utils
+from typing import Callable, Iterable, List, NamedTuple, Optional, Dict, Any, Union
+from hammer_logging import HammerVLSIFileLogger, HammerVLSILogging, HammerVLSILoggingContext
+from hammer_config import HammerDatabase
+from functools import reduce
+
+class HammerSubmitCommand:
+
+    @abstractmethod
+    def submit(self, args: List[str], env: Dict[str,str], logger: HammerVLSILoggingContext, cwd: str = None) -> str:
+        """
+        Submit the job to the job submission system. This function MUST block until the command is complete.
+
+        :param args: Command-line to run; each item in the list is one token. The first token should be the command to run.
+        :param env: The environment variables to set for the command
+        :param logger: The logging context
+        :param cwd: Working directory (leave as None to use the current working directory).
+        :return: The command output
+        """
+        pass
+
+    @abstractmethod
+    def read_settings(self, settings: Dict[str, Any], tool_namespace: str) -> None:
+        """
+        Read the settings object (a Dict[str, Any]) into meaningful class variables
+
+        :param settings: A Dict[str, Any] comprising the settings for this command
+        :param tool_namespace: The namespace for the tool (useful for logging)
+        """
+        pass
+
+
+    @staticmethod
+    def get(tool_namespace: str, database: HammerDatabase) -> "HammerSubmitCommand":
+        """
+        Get a concrete instance of a HammerSubmitCommand for a tool
+
+        :param tool_namespace: The tool namespace to use when querying the HammerDatabase (e.g. "synthesis" or "par")
+        :param database: The HammerDatabase object with tool settings
+        """
+
+        submit_command_mode = database.get_setting(tool_namespace + ".submit.command", nullvalue="none")
+        # TODO This is sketchy
+        submit_command_settings = database.get_setting("vlsi.submit.settings", nullvalue=[]) + \
+            database.get_setting(tool_namespace + ".submit.settings", nullvalue=[]) # type: List[Dict[str, Dict[str, Any]]]
```

Comment:
What's exactly preventing this from working? If you think it's small I can look into it and fix it since this does look very sketchy as you point out

Reply:
Let me chat with you in person on this one.

Reply:
Opened #224

Reply:
#230